tensorboard/plugins/pr_curve/tf_pr_curve_dashboard/tf-pr-curve-card.ts

```diff
@@ -0,0 +1,160 @@
+      type: Object,  // function: string => string
+      value: () => ({scale: runsColorScale}),
+    },
+    _runColor: {
+      type: String,
+      computed: "_computeRunColor(run)",
+    },
+    requestManager: Object,
+    _canceller: {
+      type: Object,
+      value: () => new Canceller(),
+    },
+    _attached: Boolean,
+    // The value field is a function that returns a function because Polymer
+    // will actually call the value field if the field is a function. However,
+    // we actually want the value itself to be a function.
+    _xComponentsCreationMethod: {
+      type: Object,
+      value: () => (() => {
+        const scale = new Plottable.Scales.Linear();
+        return {
+          scale: scale,
+          axis: new Plottable.Axes.Numeric(scale, 'bottom'),
+          accessor: (d: PrCurvePoint) => d.recall,
+        };
+      }),
+      readOnly: true,
+    }
+  },
+  observers: [
+      "reload(run, tag)",
+      "_runsChanged(_attached, runs.*)",
+      "_setChartData(_runToData, stepCap)",
+  ],
+  _computeRunColor(run) {
+    return this._colorScaleFunction(run);
+  },
+  _serializedRunToPrCurveData(runToPrCurveData) {
+    return JSON.stringify(runToPrCurveData);
+  },
+  attached() {
+    // Defer reloading until after we're attached, because that ensures that
+    // the requestManager has been set from above. (Polymer is tricky
+    // sometimes)
+    this._attached = true;
+    this.reload();
+  },
+  reload() {
+    if (!this._attached) {
+      return;
+    }
+    this._canceller.cancelAll();
+    const router = getRouter();
+    let url = router.pluginRoute("pr_curve", "/pr_curves");
+    url += url.indexOf('?') > -1 ? '&' : '?';
+    url += `tag=${this.tag}&runs=${this.runs.join(',')}`;
+    const updateData = this._canceller.cancellable(result => {
+      if (result.cancelled) {
+        return;
+      }
+
+      // This is a mapping between run and data.
+      const runToData = result.value;
+      this.set('_runToData', runToData);
+      this.fire('data-updated', {
+        'runToData': runToData,
+      });
+    });
+    this.requestManager.request(url).then(updateData);
+  },
+  _setChartData(runToData, stepCap) {
+    if (!runToData || !_.isNumber(stepCap)) {
+      return;
+    }
+
+    _.forOwn(runToData, (entries, run) => {
+      if (!entries || !entries.length) {
+        return;
+      }
+
+      let seriesData: PrCurvePoint[] = [];
+      let entriesIndex = entries.length - 1;
+      while (entriesIndex > 0) {
+        if (entries[entriesIndex].step <= stepCap) {
+          // This is the rightmost entry closest (or equal to) the step cap.
+          break;
+        }
+        entriesIndex--;
+      }
+      const entry = entries[entriesIndex];
+      this.fire('run-step-updated', {
+        'run': run,
+        'step': entry.step,
+      });
+
+      // Reverse the values so they are plotted in order, which allows for
+      // tool tips.
+      entry.precision.reverse();
```

Comment:
Dangerous: tracing your variables, this is equivalent to `this._runToData[run].entry.precision.reverse()`, which mutates the property. Replace this paragraph with `const precision = entry.precision.slice().reverse();` and likewise for `const recall`.